prov/rstream/src/rstream_msg.c

```diff
@@ -0,0 +1,571 @@
+#include "rstream.h"
+#include<math.h>
+
+uint32_t rstream_imm_get_len(uint64_t imm_data)
+{
+    return (imm_data & RSTREAM_MR_LEN_MASK);
+}
+
+void rstream_imm_set_len(uint64_t* imm_data, uint32_t len)
+{
+    assert(len < RSTREAM_MR_MAX);
+    *imm_data = *imm_data | len;
+}
+
+uint16_t rstream_imm_get_credits(uint64_t imm_data)
+{
+    uint64_t credits = imm_data & RSTREAM_CREDIT_MASK;
+    credits = (credits >> RSTREAM_CREDIT_OFFSET);
+    assert(credits < RSTREAM_CREDITS_MAX);
+    return credits;
+}
+
+void rstream_imm_set_credits(uint64_t* imm_data, uint16_t credits)
+{
+    assert(credits < RSTREAM_CREDITS_MAX);
+    uint64_t bcredits = credits;
+    bcredits = bcredits << RSTREAM_CREDIT_OFFSET;
+    *imm_data = *imm_data | bcredits;
+}
+
+uint64_t rstream_imm_set_msg_len(uint64_t imm_data, uint32_t msg_len)
+{
+    assert(msg_len < RSTREAM_IWARP_IMM_MSG_LEN);
+    uint64_t bmsg_len = msg_len;
+    bmsg_len = bmsg_len << RSTREAM_IWARP_IMM_MR_OFFSET;
+    return imm_data | bmsg_len;
+}
+
+uint32_t rstream_imm_get_msg_len(uint64_t imm_data)
+{
+    uint64_t msg_len = imm_data & RSTREAM_IWARP_IMM_MR_MASK;
+    msg_len = (msg_len >> RSTREAM_IWARP_IMM_MR_OFFSET);
+    assert(msg_len < RSTREAM_IWARP_IMM_MSG_LEN);
+    return msg_len;
+}
+
+char *rstream_get_next_recv_buffer(struct rstream_ep *ep)
+{
+    char *base_ptr = (char *)ep->local_mr.tx.data_start + ep->local_mr.tx.size;
+    uint64_t *offset = &ep->local_mr.recv_buffer_offset;
+    const uint32_t full_size = RSTREAM_IWARP_DATA_SIZE * ep->qp_win.max_rx_credits;
+    char *buffer = base_ptr + *offset;
+    assert((void *)buffer < ep->local_mr.rx.data_start);
+    *offset = (*offset + RSTREAM_IWARP_DATA_SIZE) % full_size;
+    return buffer;
+}
```

Comment:
Its another circular buffer since they are alloc/freed in order and the full size is pre-set. Instead of storing some of these values I calculate them on the fly. Overall it boils down to calculating the offset for the pointer.